File: VB/GrowthWare.Framework.BusinessData/DataAccessLayer/Interfaces/IDRoles.vb
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces.Base
Imports GrowthWare.Framework.Model.Enumerations
Imports GrowthWare.Framework.Model.Profiles

Namespace DataAccessLayer.Interfaces
    Public Interface IDRoles
        Inherits IDDBInteraction

        Property SecurityEntitySeqId As Integer
        Property Profile As MRoleProfile

        ''' <summary>
        ''' Gets the accounts in role.
        ''' </summary>
        ''' <returns>DataTable.</returns>
        Function GetAccountsInRole() As DataTable

        ''' <summary>
        ''' Gets the accounts not in role.
        ''' </summary>
        ''' <returns>DataTable.</returns>
        Function GetAccountsNotInRole() As DataTable

        ''' <summary>
        ''' Updates all accounts for role.
        ''' </summary>
        ''' <param name="roleSeqID">The role seq ID.</param>
        ''' <param name="securityEntityID">The security entity ID.</param>
        ''' <param name="accounts">The accounts.</param>
        ''' <param name="accountSeqID">The account seq ID.</param>
        ''' <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        Function UpdateAllAccountsForRole(ByVal roleSeqId As Integer, ByVal securityEntityId As Integer, ByVal accounts() As String, ByVal accountSeqId As Integer) As Boolean

        ''' <summary>
        ''' Saves this instance.
        ''' </summary>
        Sub Save()

        ''' <summary>
        ''' Searches the specified search criteria.
        ''' </summary>
        ''' <param name="searchCriteria">The search criteria.</param>
        ''' <returns>DataTable.</returns>
        Function Search(ByRef searchCriteria As MSearchCriteria) As DataTable

        ''' <summary>
        ''' Deletes the role.
        ''' </summary>
        Sub DeleteRole()

        ''' <summary>
        ''' Gets the profile data.
        ''' </summary>
        ''' <returns>DataRow.</returns>
        Function GetProfileData() As DataRow

        ''' <summary>
        ''' Gets the roles by security entity.
        ''' </summary>
        ''' <returns>DataTable.</returns>
        Function GetRolesBySecurityEntity() As DataTable
    End Interface
End Namespace

File: CS/GrowthWare.Framework.BusinessData/DataAccessLayer/Interfaces/IDRoles.cs
using GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces.Base;
using GrowthWare.Framework.Model.Profiles;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
{
    interface IDRoles : IDDBInteraction
    {
        /// <summary>
        /// Gets or sets the security entity seq ID.
        /// </summary>
        /// <value>The security entity seq ID.</value>
        int SecurityEntitySeqID { get; set; }

        /// <summary>
        /// Gets or sets the profile.
        /// </summary>
        /// <value>The profile.</value>
        MRoleProfile Profile { get; set; }

        /// <summary>
        /// Gets the accounts in role.
        /// </summary>
        /// <returns>DataTable.</returns>
        DataTable AccountsInRole();

        /// <summary>
        /// Gets the accounts not in role.
        /// </summary>
        /// <returns>DataTable.</returns>
        DataTable AccountsNotInRole();

        /// <summary>
        /// Updates all accounts for role.
        /// </summary>
        /// <param name="RoleSeqID">The role seq ID.</param>
        /// <param name="SecurityEntityID">The security entity ID.</param>
        /// <param name="Accounts">The accounts.</param>
        /// <param name="AccountSeqID">The account seq ID.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        bool UpdateAllAccountsForRole(int RoleSeqID, int SecurityEntityID, string[] Accounts, int AccountSeqID);

        /// <summary>
        /// Saves this instance.
        /// </summary>
        void Save();

        /// <summary>
        /// Gets a subset of information from the database 
        /// </summary>
        /// <param name="searchCriteria"></param>
        /// <returns></returns>
        DataTable Search(MSearchCriteria searchCriteria);

        /// <summary>
        /// Deletes the role.
        /// </summary>
        void DeleteRole();

        /// <summary>
        /// Gets the profile data.
        /// </summary>
        /// <returns>DataRow.</returns>
        DataRow ProfileData();

        /// <summary>
        /// Gets the roles by BU.
        /// </summary>
        /// <returns>DataTable.</returns>
        DataTable RolesBySecurityEntity();
    }
}
